### PR TITLE
[WIP] CancellationTokenSource implementation

### LINKED ISSUE #709: Use CancellationTokenSource.CancelAfter instead of a Timeout task
We could use [CancellationTokenSource.CancelAfter](https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtokensource.cancelafter?view=netframework-4.7.2) to timeout tasks

### PR DESCRIPTION
**Pull request by nessross77**
closes #709

### CODE REVIEW

File: lib/PuppeteerSharp/FrameManager.cs [hunk 1]
```diff
@@ -67,25 +67,21 @@ public async Task<Response> NavigateFrameAsync(Frame frame, string url, Navigati
             var watcher = new NavigatorWatcher(_client, this, frame, _networkManager, timeout, options);
 
             var navigateTask = NavigateAsync(_client, url, referrer, frame.Id);
-            await Task.WhenAny(
-                watcher.TimeoutOrTerminationTask,
-                navigateTask).ConfigureAwait(false);
+            await Task.Run(() => navigateTask, watcher.TimedCancellationToken).ConfigureAwait(false);
```
Comment (kblok) [inline lib/PuppeteerSharp/FrameManager.cs]: Clever 😏

File: lib/PuppeteerSharp/FrameManager.cs [hunk 2]
```diff
@@ -67,25 +67,21 @@ public async Task<Response> NavigateFrameAsync(Frame frame, string url, Navigati
...
             }
             else
             {
-                await Task.WhenAny(
-                    watcher.TimeoutOrTerminationTask,
-                    _ensureNewDocumentNavigation ? watcher.NewDocumentNavigationTask : watcher.SameDocumentNavigationTask
-                ).ConfigureAwait(false);
 
-                if (watcher.TimeoutOrTerminationTask.IsCompleted && watcher.TimeoutOrTerminationTask.Result.IsFaulted)
+                await Task.Run(() => _ensureNewDocumentNavigation ? watcher.NewDocumentNavigationTask : watcher.SameDocumentNavigationTask, watcher.TimedCancellationToken).ConfigureAwait(false);
```
Comment (kblok) [inline lib/PuppeteerSharp/FrameManager.cs]: Maybe we can move all this ` _ensureNewDocumentNavigation ? watcher.NewDocumentNavigationTask : watcher.SameDocumentNavigationTask` to a variable, so it's more clear.

File: lib/PuppeteerSharp/Helpers/CancellationTokenSourceHelper.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+            {
+                cancellationTokenSource.CancelAfter(-1);
+            }
+            else
+            {
+                cancellationTokenSource.CancelAfter(timeout);
+            }
+
+            var cancellationToken = cancellationTokenSource.Token;
+            return cancellationToken;
```
Comment (kblok) [inline lib/PuppeteerSharp/Helpers/CancellationTokenSourceHelper.cs]: We can just return `cancellationTokenSource.Token;`

File: lib/PuppeteerSharp/Helpers/CancellationTokenSourceHelper.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        /// <summary>
+        /// Creates a CancellationToken with a timeout.
+        /// </summary>
+        /// <param name="timeout">Timeout in milliseconds.</param>
+        /// <returns>The timed cancellation token.</returns>
+        public static CancellationToken GetCancellationToken(int timeout)
+        {
+            var cancellationTokenSource = new CancellationTokenSource();
+
+            if (timeout == 0)
```
Comment (kblok) [inline lib/PuppeteerSharp/Helpers/CancellationTokenSourceHelper.cs]: Can't we do only this?
```cs
if (timeout > 0)
{
    cancellationTokenSource.CancelAfter(timeout);
}
```

File: lib/PuppeteerSharp/NavigatorWatcher.cs [hunk 5]
```diff
@@ -77,7 +78,7 @@ internal class NavigatorWatcher
         public Task<bool> SameDocumentNavigationTask => _sameDocumentNavigationTaskWrapper.Task;
         public Task<bool> NewDocumentNavigationTask => _newDocumentNavigationTaskWrapper.Task;
         public Response NavigationResponse => _navigationRequest?.Response;
-        public Task<Task> TimeoutOrTerminationTask => Task.WhenAny(_timeoutTask, _terminationTaskWrapper.Task);
+        public CancellationToken TimedCancellationToken => CancellationTokenSourceHelper.GetCancellationToken(_timeout);
```
Comment (kblok) [inline lib/PuppeteerSharp/NavigatorWatcher.cs]: Let's call it `TimeoutCancellationToken`

Comment (kblok) [reply lib/PuppeteerSharp/NavigatorWatcher.cs]: We should trigger the cancellation token here https://github.com/kblok/puppeteer-sharp/blob/a890badffa63bb8dca833e8b95005e06ed54d7ba/lib/PuppeteerSharp/NavigatorWatcher.cs#L120

File: lib/PuppeteerSharp/FrameManager.cs [hunk 6]
```diff
@@ -67,25 +67,21 @@ public async Task<Response> NavigateFrameAsync(Frame frame, string url, Navigati
...
+
+                if (watcher.TimedCancellationToken.IsCancellationRequested)
                 {
-                    exception = watcher.TimeoutOrTerminationTask.Result.Exception;
+                    exception = new TaskCanceledException("Timed cancellation token timed out.");
```
Comment (kblok) [inline lib/PuppeteerSharp/FrameManager.cs]: We can throw this exception https://github.com/kblok/puppeteer-sharp/blob/a890badffa63bb8dca833e8b95005e06ed54d7ba/lib/PuppeteerSharp/Helpers/TaskHelper.cs#L28 so we don't break compatibility

**Comment by kblok** [comment]
@nessross77 also take a look at CodeFactor errors. And also thank you for stepping in :)

**Comment by kblok** [comment]
AppVeyor got stuck

File: lib/PuppeteerSharp/Helpers/CancellationTokenSourceHelper.cs [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    /// CancellationTokenSource Helper.
+    /// </summary>
+    public static class CancellationTokenSourceHelper
+    {
+        /// <summary>
+        /// Creates a CancellationToken with a timeout.
+        /// </summary>
+        /// <param name="timeout">Timeout in milliseconds.</param>
+        /// <returns>The timed cancellation token.</returns>
+        public static CancellationToken GetCancellationToken(int timeout)
```
Comment (Meir017) [inline lib/PuppeteerSharp/Helpers/CancellationTokenSourceHelper.cs]: Maybe rename to something like `GetTimingOutCancellationToken` ? Or rename the class to `TimingOutCancellationToken` and then name the method just `Get`

**Comment by Meir017** [comment]
Do we need the `Task.Run` calls?

**Comment by kblok** [comment]
I'm afraid this is not going to work. The cancellation token is a cooperative cancellation. We need to do something with that token or send that token to a lower level API.

**Comment by kblok** [comment]
Closed due to inactivity

**Pull request closed without merging**